### ISSUE TITLE
Invalid width comparison in merge-matrices, leading to fatal error for parsed matrices

### ISSUE THREAD

**Issue by cviner** [jaimicore/matrix_clustering#1]
It appears that parsing TRANSFAC matrices does not correctly exchange the number of columns for rows (since TRASFAC is a transposed version of RSAT's internal format). Therefore, when I run `merge-matrices`, during my `matrix-clustering` run, I obtain the error: "All matrices should have the same width as the first matrix". This appears to occur because the matrix (or at least its recorded numbers of rows and columns) is the transpose of what `merge-matrices` is expecting.

I have enclosed data to reproduce this error, after extracting the attached file from the ZIP and running: `$RSAT/perl-scripts/merge-matrices -i cluster_3_node_9_cat.tf -in_format tf`

[test_file.zip](https://github.com/jaimicore/matrix_clustering/files/320904/test_file.zip)

**Comment by cviner** [comment]
Further testing suggests that this is not a TRANSFAC-specific issue. Tests with the same input matrix converted to JASPAR, yielded an identical error.

**Comment by jaimicore** [comment]
I tested the matrices you sent, actually the program worked (here there is an example: http://pedagogix-tagc.univ-mrs.fr/rsat/tmp/www-data/2016/07/06/matrix-clustering_2016-07-06.184346_Aei9FO/matrix-clustering_SUMMARY.html) of course the logos are not generated, I can imagine you need to set up some options of weblogo to use the extended alphabet. 

But at least _compare-matrices_ worked well and the matrices are separated in two groups, although I cannot see the logos.

One question, why some of them have empty rows (or columns if transposed) ?

Do you have the same matrices before the clustering to trace the error? May be could be an error in a previous step, which verbosity (option **-v**) are you using?  I suggest to use -v >= 2

**Comment by cviner** [comment]
Thanks, @jaimicore!

`compare-matrices` works for me as well. The issue, however, occurs after that, during `merge-matrices`, as outlined above. Did you manage to reproduce that error?

I don't know why there are empty rows.

I will attach a set of original matrices, with the `matrix-clustering` command I used. I am using `-v 5`.

**Comment by jaimicore** [comment]
-v 5 is ok.

Yes please, the command line, the matrices and if possible, the logos, just to have a look to the motifs.

**Comment by cviner** [comment]
I think the `merge-matrices` command with the error I reported is the issue, but is not reproduced when only running `matrix-clustering` from this intermediate file. So the provided `merge-matrices` command would likely be necessary to reproduce the error. Hopefully, you will be able to reproduce it from `matrix-clustering` itself once I upload the input and its command.

**Comment by cviner** [comment]
Sure, I will upload those shortly. I don't have logos, due to the same issue you are encountering with `weblogo`.

**Comment by jaimicore** [comment]
Ok, as maybe we want to avoid (for the moment) all the vizualization stuff, there is an option in the program that produces  the non redundant set of motifs, only. 

I could modify it to add the file with clusters and their motif ID, as well. 

You can run the same command line but use this: **-return root_matrices**  , this option is faster and skip all the visual part of the program (heatmaps, trees, website)

**Comment by cviner** [comment]
Okay. I will run it that way and post the results shortly. Thanks.

**Comment by cviner** [comment]
I seem to now be unable to run `matrix-clustering` at all, with the new version of code (from a40d12d).

Upon running: `matrix-clustering -v 5 -matrix_format meme -return root_matrices -lth Ncor 0.4 -lth cor 0.6 -o  out/test -matrix MEME_motifs_for_RSAT-FEZF2-6135.UNMOD.meme`

I obtain:

```
; 2016-07-06.142033 Checking parameter values
; INFO  Labels displayed in logo trees:     id,name
; INFO  Current directory   /mnt/work1/users/home2/cviner/cytomod/experiments/linked-2016-06-15-RSAT_matrix-clustering-central_peaks-new_scripts_provided/2016-06-17-new_merge_matrix_bug_posted_GitHub_issue_1/new_FEZF2_matrix-clustering_test

; INFO  Creating directory  out
; INFO  Changing access mode    0777    out
; INFO  Current directory   /mnt/work1/users/home2/cviner/cytomod/experiments/linked-2016-06-15-RSAT_matrix-clustering-central_peaks-new_scripts_provided/2016-06-17-new_merge_matrix_bug_posted_GitHub_issue_1/new_FEZF2_matrix-clustering_test

; INFO  Creating directory  out/test_tables
; INFO  Changing access mode    0777    out/test_tables
; INFO  Current directory   /mnt/work1/users/home2/cviner/cytomod/experiments/linked-2016-06-15-RSAT_matrix-clustering-central_peaks-new_scripts_provided/2016-06-17-new_merge_matrix_bug_posted_GitHub_issue_1/new_FEZF2_matrix-clustering_test

; INFO  Creating directory  out/test_html
; INFO  Changing access mode    0777    out/test_html
; INFO  Current directory   /mnt/work1/users/home2/cviner/cytomod/experiments/linked-2016-06-15-RSAT_matrix-clustering-central_peaks-new_scripts_provided/2016-06-17-new_merge_matrix_bug_posted_GitHub_issue_1/new_FEZF2_matrix-clustering_test

; INFO  Creating directory  out/test_aligned_logos
; INFO  Changing access mode    0777    out/test_aligned_logos
; DEBUG Archive out/test_archive.zip
; INFO  Current directory   /mnt/work1/users/home2/cviner/cytomod/experiments/linked-2016-06-15-RSAT_matrix-clustering-central_peaks-new_scripts_provided/2016-06-17-new_merge_matrix_bug_posted_GitHub_issue_1/new_FEZF2_matrix-clustering_test

; INFO  Creating directory  out/test_data
; INFO  Changing access mode    0777    out/test_data
; INFO  Matrix file meme    MEME_motifs_for_RSAT-FEZF2-6135.UNMOD.meme  
; INFO  Read list of    1   matrix files
; INFO  Reading matrix from meme file version 2015      
```

Nothing happens after the last step, despite leaving it running for over two hours.

I have enclosed the matrix file (`MEME_motifs_for_RSAT-FEZF2-6135.UNMOD.meme`, ZIP'ed to `input.zip`) and would appreciate your trying to reproduce this issue.

[input.zip](https://github.com/jaimicore/matrix_clustering/files/350781/input.zip)

**Comment by jaimicore** [comment]
You forget the matrix 'alias':

matrix-clustering -v 5 -matrix_format meme -return root_matrices -lth Ncor 0.4 -lth cor 0.6 -o out/test -matrix **Coby_meme** MEME_motifs_for_RSAT-FEZF2-6135.UNMOD.meme

I tested using more stringent thresholds: matrix-clustering -v 2 -matrix_format meme -return root_matrices -lth Ncor 0.5 -lth cor 0.7 -o out/test -matrix meme MEME_motifs_for_RSAT-FEZF2-6135.UNMOD.meme -hclust_method average -calc sum -metric_build_tree Ncor

For clusters whose matrices have the same width, or singletons, the program can produce their root motifs, but when it's necessary to add empty columns to align them, the program fails.

I guess the problem comes from the matrix reader function, I could work with @jvanheld to generalize the alphabet used in the reader function, thus in theory any letter could be used.

I tried to convert the Meme format to Transfac format using this command, actually it works but the consensus are not the adequate: **convert-matrix  -from meme -to transfac -i MEME_motifs_for_RSAT-FEZF2-6135.UNMOD.meme -pseudo 1 -multiply 1 -decimals 1 -perm 0 -bg_pseudo 0.01 -return counts,consensus** 

Let me discuss with him, I will let you know.

Thank you for the feedback

**Comment by cviner** [comment]
Thanks for catching that. I had that in previous runs and that explains why this was not working. It would be useful if `matrix-clustering` could fail-fast when that argument is missing.

I appreciate your looking into this and look forward to hearing about the generalization of the reader.

**Comment by jaimicore** [comment]
Yes, totally agree, as this was one the most recent changes I did; You are the first in discover this problem.

I will fix it and then I will update the program.

**Comment by cviner** [comment]
Great; thank you!

When you are updating it, it would also be great if you could change the shebang of `matrix-clustering` to: `#!/usr/bin/env perl`. You could then add `use warnings;`, instead of the `-w`. That would prevent my needing to change it and allow others to run it regardless of the location of their Perl binary.

**Comment by cviner** [comment]
@jaimicore: I was wondering if you had an update on this.

**Comment by cviner** [comment]
@jaimicore I think this was resolved a while ago? If not, it'd be great if the package could be updated with this and a new (full RSAT) release prepared soon. We're working on finally finishing our expanded alphabet project and people will likely eventually try expanded-alphabet matrix clustering, as we emphasize it quite a bit. I'm writing it up now, and of course citing this work and acknowledging your help—thanks again!
